### PR TITLE
Client side LOD downsampling for interactive tools

### LINKED ISSUE #1519: Add simple LOD downsampling on the client side
Allow tools that will update ranges to hint that they are in the middle of an interactive update. Then renderers can (optionally) use this information together with some heuristics about the number of data points to reduce the number of points actually drawn during the operation.

### PR DESCRIPTION
**Pull request by bryevdv**
issues: closes #1519 

This PR adds some simple general level-of-detail downsampling to `GlyphRenderer`. It works behaving interactive tools set a timestamp on their plot, allowing the plot to decide to set an `interactive` flag on its view that other renderers can use as a hint. Currently works with 1D columns, e.g. color scatter pans/zooms smoothly with 60k points, and line.py with 100k points. Much more that that and we run into browser stack size limitations with out current render mapping/transform code path.  

Implementation for 2D sources (patches, multi_line) should be straightforward. 3D sources (images) will require a bit more thought, or just have it disabled for them.

### CODE REVIEW

**Comment by bryevdv** [comment]
@bokeh/core  This i marked WIP but I would like to start a discussion around a few additional options. 

* Hinting is triggered unconditionally if the interactive tool was used within a certain time. Should we try to be more sophisticated and trigger when the overall render performances dips below some threshold? Then an issue is potential flip-flopping between modes. We'd probably have to average the last several render times to avoid this. 

* Decimation is a simple constant factor (could easily be made configurable). Should this also try to be more sophisticated?

* Same comment about the interactive timing triggers. 

Basically, downsampling happens at the individual glyph level. The main question is do we try to do any sort of global optimizations?

**Comment by fpliger** [comment]
Cool! From what I see it also optimizes ranges re-calculations, right?

> Hinting is triggered unconditionally if the interactive tool was used within a certain time. Should we try to be more sophisticated and trigger when the overall render performances dips below some threshold? 
Then an issue is potential flip-flopping between modes. We'd probably have to average the last several render times to avoid this.

I think it's a good start. Being more sophisticated may be a nice improvement though because degradation may vary significantly depending on the combination number of points and how the user is "playing hard" with the tools. 
If I'm not missing something the rendering time tracking is being measured in a much more finer grain so I'd be +1 on using it to clerverly downsample when needed. 
That said it may be something that can come after, in a separate PR?

> Decimation is a simple constant factor (could easily be made configurable). Should this also try to be more sophisticated?

I'd be explicit here and let user specify a factor (at least for now).

> Same comment about the interactive timing triggers.

ditto

> Basically, downsampling happens at the individual glyph level. The main question is do we try to do any sort of global optimizations?

for instance?

**Comment by bryevdv** [comment]
@fpliger I had meant for this PR to based off another, but it was diffing against master. Now that the other PR is merged this one looks much smaller. By globaly optimizations I just meant responding to the overall render loop times at the plot level, as opposed to having each renderer just use its own local information about its recent render call times.

**Comment by bryevdv** [comment]
@bokeh/core this is ready for review I would like to merge it this week.

**Comment by fpliger** [comment]
LGTM
